### DISCUSSION TITLE
Graph Rewriting with E-Graphs

### DISCUSSION THREAD

**Discussion by cat-state** [egraphs-good/egg#180] [E-graph theory]
Is it possible to represent graph rewriting with e-graphs?
I can understand how to represent regular string and tree based languages, but I'm not sure how one would represent a graph terms within an e-graph.

**Comment by mwillsey** [comment]
Yes it should be possible, especially with the recent addition of multipatterns. The difficulty is representing DAGs (or even graphs with cycles) in terms _outside_ the e-graph; inside the e-graph should work just fine. For example, you can make a cycle by merging things with a "nonce" term (pseudocode below): 
```rust
let class = egraph.add(0 + x);
egraph.union(class, x);
```

**Comment by Bastacyclop** [comment]
It could be useful if multi-root DAGs were supported by the library APIs, and not just expressions, for example:
-  [add_expr](https://docs.rs/egg/0.9.0/egg/struct.EGraph.html#method.add_expr) doesn't work for DAGs, see https://github.com/egraphs-good/egg/pull/128#issuecomment-1092891315
- as a result, same for [explain_equivalence](https://docs.rs/egg/0.9.0/egg/struct.EGraph.html#method.explain_equivalence)

Although there's always the workaround to collect all roots into one by adding a special node which avoids making the API more complex.

**Comment by phannebohm** [comment]
Some time ago I read [this paper](https://dl.acm.org/doi/10.1145/3122955.3122956) mentioning an algebra of graphs. It's probably not very useful for graphs beyond a certain size, but maybe it's an interesting object of study.
